### ISSUE TITLE
Properly model Elements and Isotopes

### ISSUE THREAD

**Issue by hroest** [OpenMS/OpenMS#5905]
# Problem

Currently both individual elements as well as Isotopes are modelled using `Element` which is a hacky conceptually, this implies that "an isotope  is the same as an element":

```
>>> ElementDB().getElement("O")
<pyopenms.pyopenms_2.Element object at 0x7f235032b670>
>>> ElementDB().getElement("(17)O")
<pyopenms.pyopenms_2.Element object at 0x7f235032b2b0>
```

with some strange side effects, eg (17)O is treated as a completely different Element than O and has its own isotope distribution and there is no link back from (17)O to the oxygen element (especially as we dont know that this is an isotope and not an element, we can use the atomic number though)

```
>>> ElementDB().getElement("(17)O").getIsotopeDistribution().getContainer()
[<pyopenms.pyopenms_4.Peak1D object at 0x7f235032b990>]
>>> ElementDB().getElement("O").getIsotopeDistribution().getContainer()
[<pyopenms.pyopenms_4.Peak1D object at 0x7f2370355090>, <pyopenms.pyopenms_4.Peak1D object at 0x7f235032b950>, <pyopenms.pyopenms_4.Peak1D object at 0x7f235032b670>]
```

It also means that we cannot store additional information that is isotope-specific, such as number of neutrons, half lives for radioactive elements, decay modes etc.

# Solutions

The main solution would be to add another class `Isotope` to properly model isotopes, however there are a few issues with this. We would then model the relationship with a `std::vector<Isotope>` as a member of `Element`. 
 
The main problem with this is that 

(i) many parts of OpenMS rely on `ElementDB().getElement("O").getIsotopeDistribution()` which should be reasonably fast and 
(ii) `ElementDB().getElement("(17)O")` should produce something that is of type `Element` since this can be used inside an `EmpiricalFormula`. 

Therefore the easiest solution would be to model `Isotope` by inheriting from `Element` which produces a slightly wrong modelling that "Isotope is an Element" which is not quite true, but it would be easiest and least intrusive. In some sense it is also not completely false as an isotope is a special case of an element (a realization of one specific isotope instead of a mixture). We could add a new function `ElementDB::getIsotope`  which would work for `(17)O` but throw for `O` and would specifically return an Isotope object while `ElementDB::getElement` would return a down-casted version to `Element`, which would keep existing functionality working. Secondly we would have to either *cache* or compute on the fly the result of `getIsotopeDistribution()` for each `Element` since we would still need that; both of which have issues of complexity or performance.

```
class Isotope : public Element
{
  double abundance; // has to consistent with getIsotopeDistribution()
  double half_life; // in seconds
  int neutrons;
  DecayMode decay_mode; // alpha, beta, proton, etc
  bool isIsotope() {return true;} // to allow a user to cast Element -> Isotope
  bool isStable();
}
```

which would allow 
```
const Isotope* iso = element_db.getIsotope("(17)O");
const Element* ele = element_db.getElement(iso.getAtomicNumber);
std::vector<const Isotope*> isotopes = ele.getIsotopes(); 
```

The other solution would be have two classes without inheritance: `Element` and `Isotope` which would be quite complicated. We would have to modify `EmpiricalFormula` parsing to identify isotopes, replace the single `std::map<const Element*, SignedSize>` with two maps, one for `Element` and one for `Isotope` and make sure we always iterate over both when calculating anything.

**Comment by timosachsenberg** [comment]
Thanks for the nice analysis. Separating them would indeed be cleaner but I see the problems you mentioned. Would it help much if empirical formula would expect all Isotopes instead of Element? E.g. don’t distinguish between element and isotope but just treat C as an alias for the (12)C isotope?

**Comment by hroest** [comment]
@timosachsenberg no, I think that would not work because then Glucose would be `C6H12O6` and with this `EmpiricalFormula` would assume *all* carbons are C12 and there would be no isotopic peak any more and you only have monoisotopic signal. So this would not work

**Comment by jpfeuffer** [comment]
Hmm.. for me the first thing that came to my mind is:

'Element has Isotopes'

So an Element holds a vector of Isotopes. And you access an Isotope by ElementDB.getElement("C").getIsotope(13) or (2)? I don't think parsing gets much more complicated. I assume there is already code considering the additional parentheses for an isotope.

And an EmpiricalFormula would hold a vector/map of variant that is either an Element or an Isotope. Or even just Isotopes, where an Element decomposes into its single Isotopes. Or two vectors, one of each.

**Comment by timosachsenberg** [comment]
For reference:
cdk models Element and Isotope as distinct types
https://cdk.github.io/cdk/1.5/docs/api/org/openscience/cdk/interfaces/IIsotope.html
https://cdk.github.io/cdk/1.5/docs/api/org/openscience/cdk/config/IsotopeFactory.html (similar to our ElementsDB)

**Comment by jpfeuffer** [comment]
But it feels like they don't support/have the need for non-integer/average isotope distributions inside a molecular formula. Like what we try with the Element class.
